app/backend/src/couchers/interceptors.py

```diff
@@ -34,26 +70,43 @@ class AuthValidatorInterceptor(grpc.ServerInterceptor):
     terminates the call with an UNAUTHENTICATED error code.
     """
 
-    def __init__(self, get_session_for_token, allow_jailed=True):
-        self._get_session_for_token = get_session_for_token
-        self._allow_jailed = allow_jailed
+    def __init__(self):
+        self._pool = get_descriptor_pool()
 
     def intercept_service(self, continuation, handler_call_details):
-        token = parse_session_cookie(dict(handler_call_details.invocation_metadata))
+        method = handler_call_details.method
+        # method is of the form "/org.couchers.api.core.API/GetUser"
+        _, service_name, method_name = method.split("/")
 
-        if not token:
-            return unauthenticated_handler()
+        service_options = self._pool.FindServiceByName(service_name).GetOptions()
+        auth_level = service_options.Extensions[annotations_pb2.auth_level]
 
-        # None or (user_id, jailed)
-        res = self._get_session_for_token(token=token)
+        # if unknown auth level, then it wasn't set and something's wrong
+        if auth_level == annotations_pb2.AUTH_LEVEL_UNKNOWN:
+            raise Exception("Service not annotated with auth_level")
 
-        if not res:
-            return unauthenticated_handler()
+        assert auth_level in [
+            annotations_pb2.AUTH_LEVEL_OPEN,
+            annotations_pb2.AUTH_LEVEL_JAILED,
+            annotations_pb2.AUTH_LEVEL_SECURE,
+        ]
+
+        token = parse_session_cookie(dict(handler_call_details.invocation_metadata))
+
+        res = _try_get_and_update_user_details(token)
```

Comment:
Since I'm allergic to "try" or "check" functions, I'd like to get this function reworked in the following ways:

1. Don't do any database lookup when using the open API yet providing a token.
2. Fail directly by returning the unauthenticated_handler() instead of proxying it first returning None from _try_get_and_update_user_details.

The easiest is probably to inline _try_get_and_update_user_details() here and try to break out some other part.

Reply:
I would like the token/user data to be validated even on open APIs, so we can make a (verified) `user_id` available in the context.

This is useful in many cases; currently for example, the bug reporter has the frontend relay the user id (in the request) to the backend which is attached to bug reports. I would much rather have this be checked so we could be sure the user wasn't trivially spoofed.

That being said, I do think the code could be reworked. I'm trying to keep this section easy to read + understand.

Reply:
It is important to know for the frontend developer which calls it needs to send the auth cookie for. Perhaps we should have another enum value for the open calls where the auth cookie makes a difference?

Reply:
As discussed in the backend meeting: I think we should send the token always when possible. It's super useful for debugging, and will be in the future in general.

I think we just need to catch the problems here in tests.

Reply:
If you stop force-pushing maybe I can help refactor the function.

Reply:
Okay! Done force pushing for a while. I just had to rebase to update for the other proto changes.

Reply:
You could merge :) but we've had this discussion before, both ways have its pros and cons.

Reply:
I'm a rebase guy 🙃

Reply:
Basically I couldn't find any good way to refactor, so I'll be happy with how it looks now.